File: Piece of cake/144. Binary Tree Preorder Traversal/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
   public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> res;
        auto visit = [&](TreeNode *cur) { res.push_back(cur->val); };
        auto cur = root;
        while (cur) {
            auto right_most = cur->left;
            if (right_most) {
                while (right_most->right && right_most->right != cur) {
                    right_most = right_most->right;
                }
                if (right_most->right == nullptr) {
                    visit(cur);
                    right_most->right = cur;
                    cur = cur->left;
                } else {
                    cur = cur->right;
                    right_most->right = nullptr;
                }
            } else {
                visit(cur);
                cur = cur->right;
            }
        }
        return res;
    }
};

int main() { return 0; }

File: Piece of cake/144. Binary Tree Preorder Traversal/main.rs
#![allow(dead_code)]
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
  pub fn preorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut res = Vec::new();
    let mut visit = |cur: &TreeNode| {res.push(cur.val);};
    let mut cur = root.clone();
    while let Some(curr) = cur.clone() {
      let curr = curr.borrow();
      let most_right = curr.left.clone();
      if let None = most_right {
        visit(&curr);
        cur = curr.right.clone();
      } else {
        let mut most_right = most_right.unwrap();
        while most_right.borrow().right.is_some() && most_right.borrow().right != cur {
          let tmp = most_right.borrow().right.clone().unwrap();
          most_right = tmp;
        }
        if most_right.borrow().right.is_none() {
          most_right.borrow_mut().right = cur.clone();
          cur = curr.left.clone();
        } else {
          visit(&curr);
          cur = curr.right.clone();
          most_right.borrow_mut().right = None;
        }
      }
    }
    res
  }
}

fn main() {

}
struct Solution;